test/provenance/provenance_query_test.go

```diff
@@ -0,0 +1,506 @@
+}
+
+func TestProvenanceQueries(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(1)
+	setupConfig := &setup.Config{
+		NumberOfServers:     1,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+	leaderIndex := -1
+	require.Eventually(t, func() bool {
+		leaderIndex = c.AgreedLeader(t, 0)
+		return leaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	s := c.Servers[0]
+
+	// creating two databases -- db1 and db2
+	setup.CreateDatabases(t, s, []string{"db1", "db2"})
+
+	// creating two users -- alice and bob
+	aliceCert, _ := testutils.LoadTestCrypto(t, c.GetUserCertDir(), "alice")
+	bobCert, _ := testutils.LoadTestCrypto(t, c.GetUserCertDir(), "bob")
+	setup.CreateUsers(t, s, []*types.UserWrite{
+		{
+			User: &types.User{
+				Id:          "alice",
+				Certificate: aliceCert.Raw,
+				Privilege: &types.Privilege{
+					DbPermission: map[string]types.Privilege_Access{
+						"db1": types.Privilege_ReadWrite,
+						"db2": types.Privilege_ReadWrite,
+					},
+				},
+			},
+		},
+		{
+			User: &types.User{
+				Id:          "bob",
+				Certificate: bobCert.Raw,
+				Privilege: &types.Privilege{
+					DbPermission: map[string]types.Privilege_Access{
+						"db1": types.Privilege_ReadWrite,
+						"db2": types.Privilege_ReadWrite,
+					},
+				},
+			},
+		},
+	})
+
+	prepareDataForProvenanceQueries(t, s)
+
+	t.Run("Get all values", func(t *testing.T) {
+		expectedKey1Values := []*types.ValueWithMetadata{
+			{
+				Value: []byte("key1value1"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 4,
+						TxNum:    0,
+					},
+				},
+			},
+			{
+				Value: []byte("key1value2"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 5,
+						TxNum:    0,
+					},
+				},
+			},
+		}
+		resp, err := s.GetAllValues(t, "db1", "key1", "alice")
+		require.NoError(t, err)
+		require.ElementsMatch(t, expectedKey1Values, resp.GetResponse().GetValues())
+
+		expectedKey2Values := []*types.ValueWithMetadata{
+			{
+				Value: []byte("key2value1"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 4,
+						TxNum:    0,
+					},
+					AccessControl: &types.AccessControl{
+						ReadWriteUsers: map[string]bool{"alice": true},
+					},
+				},
+			},
+			{
+				Value: []byte("key2value2"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 6,
+						TxNum:    0,
+					},
+				},
+			},
+			{
+				Value: []byte("key2value3"),
+				Metadata: &types.Metadata{
+					Version: &types.Version{
+						BlockNum: 7,
+						TxNum:    0,
+					},
+				},
+			},
+		}
+		resp, err = s.GetAllValues(t, "db1", "key2", "alice")
```

Comment:
What happens when bob tries to get all values? those are not his keys right?
This is a hole in our privacy... 
We should think about what to do with this before we go to full pledged access control.

Only admins? 
Ability to disable provenance?
Or wait for provenance ACLs...

Reply:
Added scenario so that when we add ACL, a test will fail.